File: Week-1/NumberComplement-476/solution.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Number Complement.
// Memory Usage: 8.2 MB, less than 100.00% of C++ online submissions for Number Complement.

class Solution
{
public:
    int findComplement(int num)
    {
        int count = 0; //used to get the number of bits in the num
        int temp = num;

        while (temp > 0)
        {
            temp /= 2;
            count++;
        }
        num <<= 32 - count; //left shift our num by 32-count since it's a 32 bit representation;
        num = ~num;         //get the complement of num;
        num >>= 32 - count; //right shift our num by 32-count to get back to the original position

        return num;
    }
};

File: Week-1/NumberComplement-476/solution.js
// Runtime: 40 ms, faster than 99.71% of JavaScript online submissions for Number Complement.
// Memory Usage: 34.1 MB, less than 100.00% of JavaScript online submissions for Number Complement.

var findComplement = function(num) {
  var count = 0,
    temp = parseInt(num);

  while (temp > 0) {
    temp = parseInt(temp / 2);
    count++;
  }

  num <<= 32 - count;
  num = ~num;
  num >>= 32 - count;

  return num;
};
